File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/datastructures/Problem1406.java
package com.keunwon.algorithm.datastructures;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Problem1406 {

    public static void main(String[] args) throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        final String[] words = br.readLine().split("");
        final int N = Integer.parseInt(br.readLine());

        Stack<String> left = new Stack<>();
        Stack<String> right = new Stack<>();

        for (String word : words) {
            left.push(word);
        }

        for (int i = 0; i < N; i++) {
            final String[] command = br.readLine().split(" ");

            switch (command[0]) {
                case "L":
                    if (!left.isEmpty()) {
                        right.push(left.pop());
                    }
                    break;
                case "D":
                    if (!right.isEmpty()) {
                        left.push(right.pop());
                    }
                    break;
                case "B":
                    if (!left.isEmpty()) {
                        left.pop();
                    }
                    break;
                case "P":
                    left.push(command[1]);
                    break;
            }
        }

        StringBuilder sb = new StringBuilder();

        while (!left.isEmpty()) {
            right.push(left.pop());
        }

        while (!right.isEmpty()) {
            sb.append(right.pop());
        }

        System.out.println(sb.toString());
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem1406.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

/**
 * <p>
 * 이름: 에디터
 * 난이도: 실버-2
 * </p>
 *
 * @see <a href="https://www.acmicpc.net/problem/1406">에디터 (실버-2)</a>
 **/
class Problem1406 {
    fun solution(init: String, commands: Array<String>): String {
        val left = Stack<Char>().apply { init.forEach { push(it) } }
        val right = Stack<Char>()

        commands.forEach { command ->
            val arr = command.split(" ")
            when (arr[0][0]) {
                'L' -> if (left.isNotEmpty()) right.push(left.pop())
                'D' -> if (right.isNotEmpty()) left.push(right.pop())
                'B' -> if (left.isNotEmpty()) left.pop()
                'P' -> left.push(arr[1][0])
            }
        }

        return buildString(left.size + right.size) {
            while (left.isNotEmpty()) right.push(left.pop())
            while (right.isNotEmpty()) append(right.pop())
        }
    }
}

fun main() {
    val br = System.`in`.bufferedReader()

    val init = br.readLine()
    val n = br.readLine().toInt()
    val commands = Array(n) { br.readLine() }
    Problem1406().solution(init, commands).also { println(it) }
}
